pkg/identity/cache/local.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"fmt"
+
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/kvstore/allocator"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+type localIdentityCache struct {
+	mutex               lock.RWMutex
+	identitiesByID      map[identity.NumericIdentity]*identity.Identity
+	identitiesByLabels  map[string]*identity.Identity
+	nextNumericIdentity identity.NumericIdentity
+	minID               identity.NumericIdentity
+	maxID               identity.NumericIdentity
+	events              allocator.AllocatorEventChan
+}
+
+func newLocalIdentityCache(minID, maxID identity.NumericIdentity, events allocator.AllocatorEventChan) *localIdentityCache {
+	return &localIdentityCache{
+		identitiesByID:      map[identity.NumericIdentity]*identity.Identity{},
+		identitiesByLabels:  map[string]*identity.Identity{},
+		nextNumericIdentity: minID,
+		minID:               minID,
+		maxID:               maxID,
+		events:              events,
+	}
+}
+
+func (l *localIdentityCache) bumpNextNumericIdentity() {
+	if l.nextNumericIdentity == l.maxID {
+		l.nextNumericIdentity = l.minID
+	} else {
+		l.nextNumericIdentity++
+	}
+}
+
+// allocateNumericIdentity returns the next available numeric identity or an error
+func (l *localIdentityCache) allocateNumericIdentity() (identity.NumericIdentity, error) {
+	firstID := l.nextNumericIdentity
+	for {
+		if _, ok := l.identitiesByID[l.nextNumericIdentity]; !ok {
+			id := l.nextNumericIdentity
+			l.bumpNextNumericIdentity()
+			return id, nil
+		}
+
+		l.bumpNextNumericIdentity()
+		if l.nextNumericIdentity == firstID {
```

Comment:
Could `l.nextNumericIdentity` be compared to `l.minID` to indicate that there was an overflow? Or do we want to iterate over all possible identities (in which case this is fine).

Reply:
We want to iterate over all possible identities and allow reuse after the initial pass over (which is already highly unlikely) given the total space of 24bits.